dace/transformation/dataflow/map_distribution.py

```diff
@@ -0,0 +1,940 @@
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+""" Contains classes that distribute Map computations """
+
+from dace.sdfg.utils import consolidate_edges
+from copy import deepcopy
+from numbers import Number
+from typing import Dict, List
+import dace
+import sympy
+from dace import data, dtypes, registry, subsets, symbolic
+from dace.sdfg import nodes
+from dace.sdfg import utils as sdutil
+from dace.sdfg.graph import OrderedMultiDiConnectorGraph
+from dace.transformation import transformation as pm
+from dace.transformation.subgraph.helpers import subgraph_from_maps
+from functools import reduce
+
+
+@registry.autoregister_params(singlestate=True)
+class ElementWiseArrayOperation(pm.Transformation):
+    """ Distributes element-wise array operations.
+    """
+
+    _map_entry = pm.PatternNode(nodes.MapEntry)
+
+    @staticmethod
+    def expressions():
+        return [sdutil.node_path_graph(ElementWiseArrayOperation._map_entry)]
+
+    @staticmethod
+    def can_be_applied(graph: dace.SDFGState,
+                       candidate: Dict[pm.PatternNode, int],
+                       expr_index: int,
+                       sdfg: dace.SDFG,
+                       strict: bool = False):
+
+        map_entry = graph.node(candidate[ElementWiseArrayOperation._map_entry])
+        map_exit = graph.exit_node(map_entry)
+        params = [dace.symbol(p) for p in map_entry.map.params]
+
+        if "commsize" in map_entry.map.range.free_symbols:
+            return False
+        if "Px" in map_entry.map.range.free_symbols:
+            return False
+        if "Py" in map_entry.map.range.free_symbols:
+            return False
+
+        # If the map iterators are used in the code of a Tasklet,
+        # then we cannot flatten them (currently).
```

Comment:
Doesn't `subgraph.replace` take care of replacing the iterators inside the tasklets too?

Reply:
Probably. I just want to only support clean dataflow and not stuff like dappy tasklets or tasklets with custom code at the moment.